README
Modified DBT-1 for Postgres-XC test and evaluation

This folder contains DBT-1 material to test and evaluate Postgres-XC.

The following modifications were made.

1. PostgreSQL interface was changed from ODBC to libpq.   It allows 
   much more workload to apply.

2. Added housekeeping feature of ORDER table.   Original DBT-1 did not 
   maintain old ORDER table.   So performance drop was observed when 
   the benchmark is done for a long period (days to weeks).   The main 
   reason is infinite growth of ORDER table.

3. Table design change.  Table design was changed to utilize Postgres-XC's 
   parallel execution feature and to avoid cross-node operations.   
   For details of the change, please refer to Postgres-XC's architecture 
   document.

The original DBT-1 repository will be found at 

git://osdldbt.git.sourceforge.net/gitroot/osdldbt/osdldbt

And the summary will be at

http://osdldbt.git.sourceforge.net/git/gitweb.cgi?p=osdldbt/dbt1;a=summary

Postgres-XC development team appreciates Mr.Mark Wong who kindly approved 
to add this material to Postgres-XC releases.   We also appreciate if 
anybody take a look at these modifications and include some of them into 
the original DBT-1 repository.

Postgres-XC development team


ChangeLog
2009-06-11 13:28
Release v3.0.0
Improvement:
============
・libpqを使用した測定が実施できるように変更した。
  - これに伴い、configureのオプションを変更した。
・シーケンス番号を自動的に変更してシーケンスを作成するスクリプトを追加した。
  - scripts/pgsql/create_sequence.sh
・odbcのAUTOCOMMIT属性の設定順序を変更することで以下の新odbcバージョンに対応した。
  - psqlodbc-08.03.0400 / libiodbc-3.52.6
・マルチクライアント測定用の簡易実行ツールを作成した。
  - wrapper/*
bug fixes:
============
・非マルチスレッドセーフ関数をマルチスレッドセーフに対応するよう修正した。

200y-mm-dd hh:mi
Release v2.3
Improvement:
============
・使用するデータ型を以下のように変更した。
  - SQL_C_LONG → SQL_C_SBIGINT
  - SQL_C_ULONG → SQL_C_UBIGINT
  - SQL_INTEGER → SQL_C_SHORT
・ol_idを1からの通番に変更した。（interfaces/odbc_com/interaction_buy_confirm.c）
bug fixes:
============
・在庫量の更新計算ミスを修正した。
・c_passwdに小文字化したc_unameを使用するように変更した。（include/inc_pgsql/InsertCust.h）
・STMT_InsertCust_LastCIDのSQLを実行しないように修正した。（include/inc_pgsql/InsertCust.h）
・文字列変数の長さが不足していた問題を修正した。（interfaces/odbc_com/addToSC.c, createSC.c）
・memset()による初期化漏れを修正した。（interfaces/odbc_com/createSC.c）
・不必要な箇所でのCOMMITを削除した。（interfaces/odbc_com/interaction_admin_confirm.c）
・該当レコードが 0 のときの処理を追加。（interfaces/odbc_com/addToSC.c）
・STMT_BUYCONF_getCountryの呼び出し部分のコメントアウトを解除した。（interfaces/odbc_com/interaction_buy_confirm.c）
・c_idが未定の場合の処理を追加した。（interfaces/odbc_com/interaction_home.c）
・STMT_ORDER_DISPLAY_selODRInfoの結果が0件のときの処理を削除した。（interfaces/odbc_com/interaction_order_display.c）
・ループ回数の計算ミスを修正した。（interfaces/odbc_com/interaction_shopping_cart.c）
・createSC()呼び出し箇所を変更した。（interfaces/odbc_com/interaction_shopping_cart.c）

200y-mm-dd hh:mi
Release v2.1.1
Improvement:
============
・co_id変数をint型からshort型に変更した。
・テーブルカラムのnumeric型を下記の対応する型にそれぞれ変更した。
  - small int、int、bigint
・COPYによるデータロード前に、TRUNCATEを実施するように変更した。
・データロード後に主キーを作成するように変更した。
・DBT1で使用するリレーションにPUBLIC権限を付与していた箇所を削除した。
・外部キーをすべて削除した。

2003-02-03 14:48  jztpcw

	* doc/sapdb_dbt1_sizing.sxc: added sheet 'data file size'

2003-02-03 14:46  jztpcw

	* doc/osdl-dbt1-user-manual.sxw: reformatted by Dave on Sep/11/02,
	rewrote the instructions 'how to compile' to use autoconfig, added
	more detail explanation for build_db.sh and the scripts in
	/scripts/sapdb

2003-02-03 14:44  jztpcw

	* doc/: osdl-dbt1-architecture.sxw, osdl-dbt1-maintenance.sxw,
	osdl-dbt1-tpc.sxw: reformatted by Dave on Sep/11/02

2003-01-28 14:44  fimath

	* common/common.c: Removed a superfluous printf().

2003-01-28 13:29  fimath

	* cache/cache.c: Fixed a bug where the return from undo_digsyl()
	had to be decremented to be able to index the arrays with the
	search data correctly.

2003-01-28 11:49  fimath

	* cache/cache.c: Destroy the mutex_counter after warm up, since
	it's not needed anymore.
	
	Forgot to set cache_ready = 1 to let the cache run after warm up.

2003-01-27 15:53  fimath

	* cache/cache.c: Used pthread_join() to wait for threads to finish,
	instead of semaphore.
	
	Made the cache start its listener before initializing the database
	connections.
	
	Fixed the logic for dividing up the work when the cache is warmed
	up.
	
	Re-added the logic to parallelize the work to warm up the cache.
	
	Added the option to define the path to write output files.
	
	Defined a sanity check option (-s) to output the entire contents of
	the cache when it warms up, and how the division of labor is done
	for the parallelized warm up.
	
	Cache requests are not serviced until the cache is warm.
	
	Added various output messages to display the cache progress in
	warming up.
	
	Applied a consistent style in various parts of the code.

2003-01-27 15:48  fimath

	* include/cache.h: Added a copyright statement.  Changed the ifndef
	to match the filename.	Removed function prototypes for functions
	that are only used locally.

2003-01-27 15:47  fimath

	* common/common.c: Added code to support defining an output
	directory, if so desired.

2003-01-27 15:46  fimath

	* include/common.h: Defined extern output_path to support the
	option of defining an output directory for all programs.

2003-01-24 13:59  jztpcw

	* scripts/sapdb/restore_db.sh: added db_cold before changing
	parameter DATA_CACHE

2003-01-24 13:49  jztpcw

	* scripts/sapdb/backup_db.sh: added db_cold before changing
	parameter DATA_CACHE

2003-01-24 13:48  jztpcw

	* scripts/sapdb/set_param.sh: added db_cold before changing
	parameters

2003-01-24 11:28  jztpcw

	* scripts/sapdb/build_db.sh: changed usage message, changed PATH to
	DATA_PATH, otherwise system environment variable PATH is screwed up

2003-01-23 15:35  fimath

	* cache/cache.c: Reformatting some lines for consistency.  Using
	optget now.  Output some of the cache data after warm up for a
	sanity check.  Single threaded the cache warm up for correctness. 
	Using a semaphore, instead of joining with all the threads, to wait
	for the warm up to complete.

2003-01-23 15:34  fimath

	* storedproc/sapdb/getPromoImages.sql: Was told using WHERE..IN
	doesn't guarantee correctness when we're trying to query pairs. 
	Using a slow, but correct method for now.

2003-01-23 15:20  fimath

	* interfaces/odbc/odbc_interaction_search_results.c: The assignment
	of i_related and i_thumbnail variables were in the wrong order.

2003-01-21 13:32  fimath

	* datagen/main.c: By default, don't create the data files in
	parallel, it slows down for some reason.

2003-01-17 17:31  fimath

	* scripts/sapdb/: drop_dbproc.sh, drop_dbproc.sql: Moved
	drop_dbproc.sql into drop_dbprod.sh and updated the list of stored
	procedures to be dropped.

2003-01-17 17:16  jztpcw

	* scripts/sapdb/build_db.sh: moved update statistics before
	load_dbproc

cache/cache.c
/*
 *
 * This file is released under the terms of the Artistic License.  Please see
 * the file LICENSE, included in this package, for details.
 *
 * Copyright (C) 2002 Open Source Development Lab, Inc.
 * History:
 * 2002 Created by Mark Wong & Jenny Zhang &
 * Aug-2003: Rewrote parsing the command line part by Jenny Zhang
 */

#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <ctype.h>
#include <sys/types.h>
#include <signal.h>
#include <getopt.h>

#include <common.h>
#include <_socket.h>
#include <cache.h>
#include <cache_interface.h>
#include <db.h>

struct search_results *author_results_table;
struct search_results *title_results_table;
void *warm_up_cache(void *fd);
void *cache_thread(void *fd);
void sighandler(int signum);

pthread_mutex_t mutex_cache_server = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t mutex_counter = PTHREAD_MUTEX_INITIALIZER;
int i_id_max, a_id_max;
#ifdef ODBC
	char sname2[32] = "DBT1";
	char uname2[32] = "pgsql";
	char auth2[32] = "pgsql";
#endif
#ifdef LIBPQ
	char sname2[32] = "localhost";
	char dbname2[32] = "DBT1";
	char uname2[32] = "pgsql";
	char auth2[32] = "pgsql";
	char dbport2[32]="5432";
#endif
int db_thread=10;
int sanity_check = 0;
int cache_ready = 0;
int warm_up = 0;
int counter = 0;
int item_count = 1000;
int port = 9999;
int help = 0;

void *init_cache(void *data);
void *warm_up_thread(void *fd);
void status();
int undo_digsyl(char *search_string);
int usage(char *name);

int main(int argc, char *argv[])
{
	int mastersock, workersock;
	struct sockaddr_in socketaddr;
	pthread_t tid;
	pthread_t connect_thread;
	int connectioncount;
	int addrlen;
	int rec;
	struct sigaction sa;
	int c;

	setlinebuf(stdout);
	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = &sighandler;
	if (sigaction(SIGUSR1, &sa, NULL) == -1)
	{
		printf("can not register signal handler to SIGUSR1\n");
		return -1;
	}
	if (sigaction(SIGUSR2, &sa, NULL) == -1)
	{
		printf("can not register signal handler to SIGUSR2\n");
		return -1;
	}
	if (sigaction(SIGSEGV, &sa, NULL) == -1)
	{
		printf("can not register signal handler to SIGSEGV\n");
		return -1;
	}

	if (argc < 1)
	{
		return usage(argv[0]);
	}

	while (1)
	{
		static struct option long_options[] = {
			{ "host", required_argument, 0, 0 },
			{ "dbname", required_argument, 0, 0 },
			{ "dbnodename", required_argument, 0, 0 },
			{ "username", required_argument, 0, 0 },
			{ "password", required_argument, 0, 0 },
			{ "output_path", required_argument, 0, 0 },
			{ "port", required_argument, 0, 0 },
			{ "db_connection", required_argument, 0, 0 },
			{ "item_count", required_argument, 0, 0 },
			{ "help", no_argument, &help, 1},
			{ "sanity_check", no_argument, &sanity_check, 1},
			{ 0, 0, 0, 0 }
		};
		int option_index = 0;
		c = getopt_long_only(argc, argv, "", long_options, &option_index);
		if (c == -1)
		{
			 break;
		}

		switch (c)
        	{
                case 0:
			if (long_options[option_index].flag != 0)
			{
				break;
			}
			if (strcmp(long_options[option_index].name, "host") == 0)
			{
				strcpy(sname2, optarg);
			}
			else if (strcmp(long_options[option_index].name, "help") == 0)
			{
				break;
			}
			else if (strcmp(long_options[option_index].name, "dbnodename") == 0)
			{
				strcpy(sname2, optarg);
			}
#ifdef LIBPQ
           		else if (strcmp(long_options[option_index].name, "dbname") == 0)
			{
				strcpy(dbname2, optarg);
			}
#endif
            		else if (strcmp(long_options[option_index].name, "username") == 0)
			{
				strcpy(uname2, optarg);
			}
			else if (strcmp(long_options[option_index].name, "password") == 0)
			{
				strcpy(auth2, optarg);
			}
			else if (strcmp(long_options[option_index].name, "output_path") == 0)
			{
				strcpy(output_path, optarg);
			}
			else if (strcmp(long_options[option_index].name, "port") == 0)
			{
				port = atoi(optarg);
			}
			else if (strcmp(long_options[option_index].name, "db_connection") == 0)
			{
				db_thread = atoi(optarg);
			}
			else if (strcmp(long_options[option_index].name, "item_count") == 0)
			{
				item_count = atoi(optarg);
			}
			break;
		default:
			printf ("?? getopt returned character code 0%o ??\n", c);
			exit(1);
		}
	}

	if ( help == 1)
	{
		return usage(argv[0]);
	}

	a_id_max = item_count / 10;
	i_id_max = item_count / 5;

	init_common();
	
	/* Start the listenter. */
	mastersock = _server_init_socket(port);
	if (mastersock < 0)
	{
		printf("init cache server master socket failed\n");
		return -1;
	}

	/* Initialize the cache. */
	if (pthread_create(&tid, NULL, init_cache, NULL) != 0)
	{
		printf("pthread_create for init_cache failed\n");
		return -1;
	}

	printf("The cache server is active...\n");

	connectioncount = 0;
	while (1) {
		addrlen = sizeof(socketaddr);
		pthread_mutex_lock(&mutex_cache_server);
		workersock = accept(mastersock, (struct sockaddr *) &socketaddr,
			(socklen_t *)&addrlen);
		if (workersock < 0) {
			LOG_ERROR_MESSAGE("accept couldn't open worker socket, errno %d",
			errno);
		}

		connectioncount++;
		if ((rec = pthread_create(&connect_thread, NULL, cache_thread,
			&workersock)) != 0)    
		{
			LOG_ERROR_MESSAGE("pthread_create failed, rec=%d", rec);
			return -1;
		}
	}
	printf("Shutting down.\n");

	close(mastersock);
 
	return 0;
	
}

void sighandler(int signum)
{
	if (signum == SIGUSR1)
		printf("connect to database failed\n");
	else if (signum == SIGUSR2)
		printf("warm up threads failed\n");
	else if (signum == SIGSEGV)
		printf("memory violation\n");
	printf("please check error.log\n");
	_exit(-1);
}

/*
 * Allocate the results tables, initialize odbc environment, and warm up cache.
 */
void *init_cache(void *data)
{
	int i, j;
	pthread_t *warm_up_thread_id;
	struct table_range *range;
	int author_step, title_step;

#ifdef ODBC
	if (db_init(sname2, uname2, auth2) != OK)
#endif
#ifdef LIBPQ
	if (db_init(sname2, dbname2, uname2, auth2, dbport2) != OK)
#endif
	
	{
		printf("db environment initialization failed\n");
		return NULL;
	}

	author_results_table = malloc(a_id_max * sizeof(struct search_results));	
	title_results_table = malloc(i_id_max * sizeof(struct search_results));	
	for (i = 0; i < a_id_max; i++)
	{
		for(j = 0; j < PROMOTIONAL_ITEMS_MAX; j++)
		{
			author_results_table[i].i_related[j] = 0;
			author_results_table[i].i_thumbnail[j] = 0;
		}
	}
	for (i = 0; i < i_id_max; i++)
	{
		for(j = 0; j < PROMOTIONAL_ITEMS_MAX; j++)
		{
			title_results_table[i].i_related[j] = 0;
			title_results_table[i].i_thumbnail[j] = 0;
		}
	}

	warm_up_thread_id = (pthread_t *) malloc(db_thread * sizeof(pthread_t));
	range =
		(struct table_range *) malloc(db_thread * sizeof(struct table_range));

	/* Determine the range of the data that each thread will retrieve. */
	author_step = a_id_max / db_thread;
	title_step = i_id_max / db_thread;
	range[0].author_start = 0;
	range[0].author_end = author_step - 1;
	range[0].title_start = 0;
	range[0].title_end = title_step - 1;
	for (i = 1; i < db_thread; i++)
	{
		range[i].author_start = range[i - 1].author_end + 1;
		range[i].author_end = range[i].author_start + author_step - 1;
		range[i].title_start = range[i - 1].title_end + 1;
		range[i].title_end = range[i].title_start + title_step - 1;
	}
	if (db_thread > 1)
	{
		/* Set the last range's end numbers in case of rounding error. */

		range[db_thread - 1].author_end = a_id_max - 1;
		range[db_thread - 1].title_end = i_id_max - 1;
	}

	if (sanity_check == 1)
	{
		printf("thread      a_id range      i_id range\n");
		for (i = 0; i < db_thread; i++)
		{
			printf("%6d  %5d to %5d  %5d to %5d\n", i,
				range[i].author_start + 1, range[i].author_end + 1,
				range[i].title_start + 1, range[i].title_end + 1);
		}
	}

	/* create threads to fill out the results tables */
	for (i = 0; i < db_thread; i++)
	{
		if (pthread_create(&warm_up_thread_id[i], NULL, warm_up_cache,
			(void *) &range[i]) != 0)
		{
			printf("warm_up thread create failed");
			return NULL;
		}
		sleep(1);
	}
	printf("%d rows to retrieve.\n", a_id_max + i_id_max);
	warm_up = 1;

	/* Wait for all threads to finish warming up the cache. */
	for (i = 0; i < db_thread; i++)
	{
		pthread_join(warm_up_thread_id[i], NULL);
	}

	printf("Cache is warm.\n");
	fflush(stdout);

	pthread_mutex_destroy(&mutex_counter);
	cache_ready = 1;

	if (sanity_check == 1)
	{
		printf("Data sanity check.\n");

		/* Check results for an a_lname search. */
		printf("Results for a_lname search:\n");
		printf("%10s %3s %10s %10s %10s %10s %10s %10s %10s %10s %10s %10s\n",
			"a_id", "#", "i_r1", "i_t1", "i_r2", "i_t2", "i_r3", "i_t3",
			"i_r4", "i_t4", "i_r5", "i_t5");
		for (i = 0; i < a_id_max; i++)
		{
			printf("%10d %3d", i + 1, author_results_table[i].record_number);
			for(j = 0; j < PROMOTIONAL_ITEMS_MAX; j++)
			{
				printf(" %10lld %10lld",
					author_results_table[i].i_related[j],
					author_results_table[i].i_thumbnail[j]);
			}
			printf("\n");

			printf("\t%10s %20s %20s %60s\n", "i_id", "a_fname", "a_lname",
				"i_title");
			for (j = 0; j < author_results_table[i].record_number; j++)
			{
				printf("\t%10lld %20s %20s %60s\n",
					author_results_table[i].search_results[j].i_id,
					author_results_table[i].search_results[j].a_fname,
					author_results_table[i].search_results[j].a_lname,
					author_results_table[i].search_results[j].i_title);
			}
			printf("\n");
		}

		/* Check results for an i_title search. */
		printf("Results for i_title search:\n");
		printf("%10s %3s %10s %10s %10s %10s %10s %10s %10s %10s %10s %10s\n",
			"a_id", "#", "i_r1", "i_t1", "i_r2", "i_t2", "i_r3", "i_t3",
			"i_r4", "i_t4", "i_r5", "i_t5");
		for (i = 0; i < i_id_max; i++)
		{
			printf("%10d %3d", i + 1, title_results_table[i].record_number);
			for(j = 0; j < PROMOTIONAL_ITEMS_MAX; j++)
			{
				printf(" %10lld %10lld",
					title_results_table[i].i_related[j],
					title_results_table[i].i_thumbnail[j]);
			}
			printf("\n");

			printf("\t%10s %20s %20s %60s\n", "i_id", "a_fname", "a_lname",
				"i_title");
			for (j = 0; j < title_results_table[i].record_number; j++)
			{
				printf("\t%10lld %20s %20s %60s\n",
					title_results_table[i].search_results[j].i_id,
					title_results_table[i].search_results[j].a_fname,
					title_results_table[i].search_results[j].a_lname,
					title_results_table[i].search_results[j].i_title);
			}
			printf("\n");
		}
		printf("Data sanity check complete.\n");
	}

	return NULL;
}

/* thread filling out results tables */
void *warm_up_cache(void *fd)
{
	struct table_range *range;
	int rc, i, j, k;
	struct db_context_t dbc;
	union interaction_data_t data;

	range = (struct table_range *)(fd);

	/* Must initialize random number generator for Promotional Processing. */
	srand(time(NULL) + pthread_self());

	rc = db_connect(&dbc);
	if (rc == ERROR)
	{
		LOG_ERROR_MESSAGE("could not connect to the database");
		kill(0, SIGUSR1);
		pthread_exit(NULL);
	}

	while (warm_up == 0)
	{
		sleep(2);
	}

	/* do search_by_author*/
	data.search_results_data.search_type = SEARCH_AUTHOR;
	for (i = range->author_start; i <= range->author_end; i++)
	{
		digsyl2(data.search_results_data.search_string,
			(long long) i + 1, (long long) 7);
		rc = process_interaction(SEARCH_RESULTS, &dbc, &data);
		if (rc == ERROR)
		{
			kill(0, SIGUSR2);
			pthread_exit(NULL);
		}

		/* copy the results out */
		//2005/08/05
		//Add limit of SEARCH_RESULT_ITEMS_MAX
		if( data.search_results_data.items > SEARCH_RESULT_ITEMS_MAX ){
			data.search_results_data.items = SEARCH_RESULT_ITEMS_MAX ;
		}
		author_results_table[i].record_number =
			data.search_results_data.items;
			
		for (j = 0; j < data.search_results_data.items; j++)
		{
			author_results_table[i].search_results[ j ].i_id =
				data.search_results_data.results_data[ j ].i_id ;
			strcpy(author_results_table[i].search_results[j].i_title,
				data.search_results_data.results_data[j].i_title);
			strcpy(author_results_table[i].search_results[j].a_fname,
				data.search_results_data.results_data[j].a_fname);
			strcpy(author_results_table[i].search_results[j].a_lname,
				data.search_results_data.results_data[j].a_lname);
		}
		for (k = 0; k < PROMOTIONAL_ITEMS_MAX; k++)
		{
			author_results_table[i].i_related[k] =
				data.search_results_data.pp_data.i_related[k];
			author_results_table[i].i_thumbnail[k] =
				data.search_results_data.pp_data.i_thumbnail[k];
		}
		status();
	}

	/* do search_by_title*/
	data.search_results_data.search_type = SEARCH_TITLE;
	for (i = range->title_start; i <= range->title_end; i++)
	{
		digsyl2(data.search_results_data.search_string,
			(long long) i + 1, (long long) 7);
		rc = process_interaction(SEARCH_RESULTS, &dbc, &data);
		if (rc == ERROR)
		{
			kill(0, SIGUSR2);
			pthread_exit(NULL);
		}

		/* copy the results out */
		//2005/08/05
		//Add limit fo 
		if( data.search_results_data.items > SEARCH_RESULT_ITEMS_MAX ){
			data.search_results_data.items = SEARCH_RESULT_ITEMS_MAX ;
		}
		title_results_table[i].record_number =
			data.search_results_data.items;

		for (j = 0; j < data.search_results_data.items; j++)
		{
			title_results_table[i].search_results[j].i_id =
				data.search_results_data.results_data[j].i_id;
			strcpy(title_results_table[i].search_results[j].i_title,
				data.search_results_data.results_data[j].i_title);
			strcpy(title_results_table[i].search_results[j].a_fname,
				data.search_results_data.results_data[j].a_fname);
			strcpy(title_results_table[i].search_results[j].a_lname,
				data.search_results_data.results_data[j].a_lname);
		}
		for (k = 0; k < PROMOTIONAL_ITEMS_MAX; k++)
		{
			title_results_table[i].i_related[k] =
				data.search_results_data.pp_data.i_related[k];
			title_results_table[i].i_thumbnail[k] =
				data.search_results_data.pp_data.i_thumbnail[k];
		}
		status();
	}
	db_disconnect(&dbc);
	return NULL;
}

void *cache_thread(void *fd)
{
	int *sf;
	int workersock;
	int rec;
	struct search_results_t search_results;
	int index, i;

	sf = (int *)(fd);
	workersock = *sf;
	pthread_mutex_unlock(&mutex_cache_server);

	while (cache_ready == 0)
	{
		sleep(1);
	}

	while (1)
	{
		/* receive search results request */
		if ((rec=receive_search_results(workersock, &search_results)) == ERROR)
		{
			LOG_ERROR_MESSAGE("receive_search_results_request failed");
			close(workersock);
			pthread_exit(NULL);
		}		
		if (rec == 0)
		{
			LOG_ERROR_MESSAGE("app server socket closed");
			close(workersock);
			pthread_exit(NULL);
		}
		
		/* look up the results*/
		index = undo_digsyl(search_results.search_string) - 1;
		if (index < 0)
		{
			bzero(&search_results, sizeof(struct search_results_t));
			LOG_ERROR_MESSAGE("undo_digsyl returned index %s", index);
		}
		else if (search_results.search_type == SEARCH_AUTHOR)
		{
			if (index >= a_id_max)
			{
				bzero(&search_results, sizeof(struct search_results_t));
				LOG_ERROR_MESSAGE("undo_digsyl returned out of range index %s", index);
			}
			/* copy out data */
			search_results.items=author_results_table[index].record_number;
			for(i=0; i<author_results_table[index].record_number; i++)
			{
				search_results.results_data[i].i_id =
					author_results_table[index].search_results[i].i_id;
				strcpy(search_results.results_data[i].i_title,
					author_results_table[index].search_results[i].i_title);
				strcpy(search_results.results_data[i].a_fname,
					author_results_table[index].search_results[i].a_fname);
				strcpy(search_results.results_data[i].a_lname,
					author_results_table[index].search_results[i].a_lname);
			}
			/* copy out promotion data */
			for (i = 0; i < PROMOTIONAL_ITEMS_MAX; i++)
			{
				search_results.pp_data.i_related[i] =
					author_results_table[index].i_related[i];
				search_results.pp_data.i_thumbnail[i] =
					author_results_table[index].i_thumbnail[i];
			}
		}
		else if (search_results.search_type == SEARCH_TITLE)
		{
			if (index >= i_id_max)
			{
				bzero(&search_results, sizeof(struct search_results_t));
				LOG_ERROR_MESSAGE("undo_digsyl returned out of range index %s", index);
			}
			search_results.items = title_results_table[index].record_number;
			/* copy out data */
			for(i = 0; i < title_results_table[index].record_number; i++)
			{
				search_results.results_data[i].i_id =
					title_results_table[index].search_results[i].i_id;
				strcpy(search_results.results_data[i].i_title,
					title_results_table[index].search_results[i].i_title);
				strcpy(search_results.results_data[i].a_fname,
					title_results_table[index].search_results[i].a_fname);
				strcpy(search_results.results_data[i].a_lname,
					title_results_table[index].search_results[i].a_lname);
			}
			/* copy out promotion data */
			for (i = 0; i < PROMOTIONAL_ITEMS_MAX; i++)
			{
				search_results.pp_data.i_related[i] =
					title_results_table[index].i_related[i];
				search_results.pp_data.i_thumbnail[i] =
					title_results_table[index].i_thumbnail[i];
			}
		}
		else
		{
			LOG_ERROR_MESSAGE("invalid search type %d",
				search_results.search_type);
			bzero(&search_results, sizeof(struct search_results_t));
		}		
		
		
		/* send results back */
		if ((rec = send_search_results(workersock, &search_results)) == ERROR)
		{
			LOG_ERROR_MESSAGE("send_search_results_response failed");
			close(workersock);
			pthread_exit(NULL);
		}		
	}
	close(workersock);
}

void status()
{
	pthread_mutex_lock(&mutex_counter);
	++counter;
	if ((counter % 100) == 0)
	{
		printf("%d rows retrieved.\n", counter);
	}
	pthread_mutex_unlock(&mutex_counter);
}

int undo_digsyl(char *search_string)
{
	int i, j, k;
	int rec, digit, base;

	j = 0;
	rec = 0;
	for(i = 13; i >= 0; i = i - 2)
	{
		if (search_string[i] == 'A') digit = 0;
		else if (search_string[i] == 'I') digit = 3;
		else if (search_string[i] == 'T') digit = 6;
		else if (search_string[i] == 'N') digit = 8;
		else if (search_string[i] == 'G')
		{
			if (search_string[i - 1] == 'O') digit = 1;
			else if (search_string[i-1] == 'N') digit = 9;
			else 
			{
				LOG_ERROR_MESSAGE("no digit for syllable %cG",
					search_string[i-1]);
				return -1;
			}
		}
		else if (search_string[i] == 'L')
		{
			if (search_string[i-1] == 'A') digit = 2;
			else if (search_string[i-1] == 'U') digit = 7;
			else 
			{
				LOG_ERROR_MESSAGE("no digit for syllable %cL",
					search_string[i-1]);
				return -1;
			}
		}
		else if (search_string[i] == 'E')
		{
			if (search_string[i - 1] == 'R') digit = 4;
			else if (search_string[i - 1] == 'S') digit = 5;
			else 
			{
				LOG_ERROR_MESSAGE("no digit for syllable %cE",
					search_string[i-1]);
				return -1;
			}
		}
		else 
		{
			LOG_ERROR_MESSAGE("no digit for syllable %c%c", search_string[i-1],					search_string[i]);
			return -1;
		}
		base = 1;
		for(k = 0; k < j; k++)
			base = base * 10;
		rec += base * digit;
		j++;
	}
	return rec;
}

int usage(char *name)
{
#ifdef ODBC
	printf("usage: %s --dbnodename <dbnodename> \n", name);
#endif
#ifdef LIBPQ
	printf("usage: %s --host <hostname> --dbname <dbname> \n", name);
#endif
	printf("--username <username> --password <password>\n");
	printf("--port <port> --db_connection <connections>\n");
	printf("--item_count <items> --output_path <output_path>\n");
	printf("if you want sanity_check, add: ");
	printf("--sanity_check\n\n");

	printf("if not defined, the default values are:\n");
#ifdef ODBC
        printf("--dbnodename %s ", sname2);
#endif
#ifdef LIBPQ
        printf("--host %s --dbname %s ", sname2, dbname2);
#endif
	printf("--username %s --password %s\n", uname2, auth2);
	printf("--port %d --db_connection %d\n", port, db_thread);
	printf("--item_count %d --output_path %s\n", item_count, output_path );
	return 1;
}
